arangosh/Benchmark/BenchmarkThread.h

```diff
@@ -271,357 +245,222 @@ class BenchmarkThread : public arangodb::Thread {
-    char const* payload = _operation->payload(&payloadLength, _threadNumber,
-                                              threadCounter, globalCounter, &mustFree);
+    
+    _requestData.clear();
+    _operation->buildRequest(_threadNumber, threadCounter, globalCounter, _requestData);
+      
+    velocypack::Slice payloadSlice = _requestData.payload.slice();
+    char const* p = nullptr;
+    size_t length = 0;
+
+    if (!payloadSlice.isNone()) {
+      if (_useVelocyPack) {
+        // send as raw velocypack
+        p = payloadSlice.startAs<char const>();
+        length = payloadSlice.byteSize();
+      } else {
+        // send as stringified JSON
+        // dump JSON string into reusable payload string
+        _payloadBuffer.clear();
+        velocypack::StringSink sink(&_payloadBuffer);
+        velocypack::Dumper dumper(&sink);
+        dumper.dump(payloadSlice);
+        p = _payloadBuffer.data();
+        length = _payloadBuffer.size();
+      }
+    }
 
+    TRI_ASSERT(p != nullptr || length == 0);
+    
     double start = TRI_microtime();
-    httpclient::SimpleHttpResult* result =
-        _httpClient->request(type, url, payload, payloadLength, _headers);
+    std::unique_ptr<httpclient::SimpleHttpResult> result(_httpClient->request(_requestData.type, _requestData.url, p, length, _headers));
     double delta = TRI_microtime() - start;
-    oneTime(delta);
-    _time += delta;
+    trackTime(delta);
 
-    if (mustFree) {
-      TRI_Free((void*)payload);
-    }
+    processResponse(result.get(), /*batch*/ false);
+    
+    _httpClient->recycleResult(std::move(result));
+  }
 
-    if (result == nullptr || !result->isComplete()) {
-      _operationsCounter->incFailures(1);
-      if (result != nullptr) {
-        _operationsCounter->incIncompleteFailures(1);
-        delete result;
-      }
-      _warningCount++;
-      if (_warningCount < MaxWarnings) {
-        LOG_TOPIC("f5982", WARN, arangodb::Logger::FIXME)
-            << "single operation failed because server did not reply";
+  void processResponse(httpclient::SimpleHttpResult const* result, bool batch) {
+    char const* type = (batch ? "batch" : "single");
+
+    if (result != nullptr && 
+        result->isComplete() &&
+        !result->wasHttpError()) {
+      
+      if (batch) {
+        // for batch requests we have to check the error header in addition
+        auto const& headers = result->getHeaderFields();
+        if (auto it = headers.find(StaticStrings::Errors); it != headers.end()) {
+          uint32_t errorCount = basics::StringUtils::uint32((*it).second);
+          if (errorCount > 0) {
+            _operationsCounter->incFailures(errorCount);
+            if (++_warningCount < maxWarnings) {
+              LOG_TOPIC("b1db5", WARN, arangodb::Logger::BENCH)
+                << type << " operation sServer side warning count: " << errorCount;
+            }
+          }
+        }
       }
+
       return;
```

Comment:
If `result != nullptr && result->isComplete() && !result->wasHttpError() && !batch` is satisfied we will return here without updating any counters - this is a change to be previous behavior.

Reply:
I think the change is valid, because counters should only increased when there are any errors (the counters are simply error counters).

For non-batch requests (i.e. `!batch`) an error was previously assumed if `result == nullptr || !result->isComplete() || result->wasHttpError()`.
In the refactored version, we are still counting in this case.

And for batch requests (.e. `batch`) this case should also be ok, because for complete HTTP responses the number of errors is only in some header, which we are still checking.

And in case `result == nullptr || !result->isComplete() || result->wasHttpError()` we are still counting everything, regardless of batch requests or single requests.

Reply:
Ah, so basically a `batch` request should never report an HTTP error?
I saw that you tried to unify the error handling for the single/batch requests, but unfortunately I still think this function does not behave like the previous code. For example, for batch requests the old code increased `_operationsCounter` by `numOperations` while now we always increment by one. Also, the old code for batch requests had an optional verbose output that is no longer available in the new code.
Can we have batch requests with a single operation? If not, I would replace the `batch` argument with a `numOperations` argument and derive `bool batch = numOperations > 1`.